### ISSUE TITLE
Service methods are not executed in transaction

### ISSUE THREAD

**Issue by alhafoudh** [feathersjs/feathers-knex#91]
### Steps to reproduce

1. Make any service method call (get, find, create, update, delete)
2. Throw an error

### Expected behavior

1. Begin transaction when starting a service method.
2. If error is thrown transaction should be rolled back.

### Actual behavior

When error is thrown, data are commited to database.

**Comment by hubgit** [comment]
I'd like to be able to able to remove associated items from multiple services at once, without leaving orphan items if one of the remove operations fails for some reason.

Might it be useful if the knex adapter provided methods for opening/closing a transaction which could be called in `before`/`after` hooks respectively? The adapter could chain any database calls while a transaction is open, then execute the transaction when it's closed.

**Comment by ekryski** [comment]
@alhafoudh when you say "throw an error", do you mean if a DB error occurs or are you referring to when you throw an error in a hook or service explicitly?

**Comment by alhafoudh** [comment]
@ekryski any error, best practice I want to exercise is to wrap whole request to transaction. I cannot do that with feathers-knex

**Comment by ekryski** [comment]
Well, if an error happens in a before or after hook then it's outside of our control I think and that part is left up to you. Inside the actual `feathers-sequelize` service adapter (ie. when it's actually writing to the DB) I think that makes sense.

Am I missing anything?

**Comment by alhafoudh** [comment]
Thats the point. Its outside of my control and that is not desirable.
I would this:

1. HTTP request comes in
2. Start DB transaction
3. before hook
4. call service method
5. after hook
6. commit transaction

At least I would like to start transaction in **before hook** and pass it along, so **service method** can perform DB calls inside that same transaction and in the end **after hook** will decide commit/rollback.

**Comment by daffl** [comment]
This should be possible with application hooks and something like

```js
app.hooks({
  before(hook) {
    return new Promise(resolve => {
      knex.transaction(trx => {
        hook.transaction = trx;

        resolve(hook);
      });
    });
  },

  after(hook) {
    return hook.transaction.commit().then(() => hook);
  },

  error(hook) {
    return hook.transaction.rolback().then(() => hook);
  }
});
```

Now someone just needs to figure out how to make the query use that transaction (it'll probably be added to `params` but I'm not sure how the service adapter would us it - just add `.transacting(params.transaction)`?)

**Comment by dschnare** [comment]
I would suggest to have the developer set `transaction` on `hook.params` and if that param is present then any DB adapter can choose to handle transactions internally. So if we had all adapters handle the `hook.params.transaction` (and obviously only using it if it's a transaction object they expect) then this would work out.

```js
// The developer's code to enable transactions for the entire application
app.hooks({
  before(hook) {
    return new Promise(resolve => {
      knex.transaction(trx => {
        hook.params.transaction = trx;
        resolve(hook);
      });
    });
  },

  after(hook) {
    return hook.params.transaction.commit().then(() => hook);
  },

  error(hook) {
    return hook.params.transaction.rolback().then(() => hook);
  }
});
```

For the `feathers-kenx` adapter that "handling" could simply be this:

```js
// Refactor createQuery() to accept db
createQuery (db, queryParams = {}) { ... }
// Refactor db() to accept a transaction
db (transaction = null) {
  const db = this.knex(this.table)
  // TODO: Check if transaction is an object knex.transacting() expects
  return transaction ? db.transacting(transaction) : db
}
// Then refactor everywhere db() and createQuery() are called to pass in the new arguments
```

The handling would likely be quite similar for other adapters. This will work nicely for a single DB adapter, but won't work for projects that need to communicate to multiple databases though. Off the top of my head I don't have a solution for that scenario, but it would be great if transactions could be started for each DB. I am working on a project currently that is communicating with multiple databases so this would be very helpful to have.

**Comment by escobar5** [comment]
I'm using the transactions hooks defined [here](https://docs.feathersjs.com/api/databases/knexjs.html#transaction-support), but the service methods are not executing in transactions, this is my code:


```
class ProvisionService {
  find(params) {}
  get(id, params) {}
  async create(data, params) {
    var company = await this.app.service('companies').create({ name: data.company_name });
   ...
    await this.app.service('users').create(users);
  }
  update(id, data, params) {}
  patch(id, data, params) {}
  remove(id, params) {}
  setup(app, path) {
    this.app = app;
  }
}
```

This are the hooks

```
import { hooks as authHooks } from 'feathers-authentication';
import { hooks as dbHooks } from 'feathers-knex';

export default {
  before: {
    all: [ dbHooks.transaction.start() ],
    find: [],
    get: [],
    create: [],
    update: [],
    patch: [],
    remove: []
  },

  after: {
    all: [ dbHooks.transaction.end() ],
    find: [],
    get: [],
    create: [],
    update: [],
    patch: [],
    remove: []
  },

  error: {
    all: [ dbHooks.transaction.rollback() ],
    find: [],
    get: [],
    create: [],
    update: [],
    patch: [],
    remove: []
  }
};

```


I call the provisioning.create from rest, and it creates the company, then an error is thrown in the creation of the users, but the company insertion is not rolled back.

**Comment by daffl** [comment]
If you are calling other services through hooks you have to pass them `hook.params.transaction` to use the same transaction as [documented here](https://docs.feathersjs.com/api/databases/knexjs.html#transaction-support).

**Comment by escobar5** [comment]
I've also tried that like this:

`await this.app.service('users').create(users, { transaction: params.transaction });`

Am I doing it wrong?

**Comment by daffl** [comment]
Looks right. Can you create a new issue with the minimum steps to reproduce.

**Comment by dash7ou** [comment]
why we need to use a transaction.start() !!!!!!!!

**Comment by KidkArolis** [comment]
Because not everything needs to be done in a transaction. You want to have fine control over when and how it's used.
